snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field
+from threading import Event, Lock
+from typing import Callable, Generic, Mapping, MutableMapping, Optional, Set, Sequence
+
+from snuba.utils.concurrent import execute
+from snuba.utils.streams.consumers.backends.abstract import ConsumerBackend
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+
+# TODO: TOffset needs to be bound as some sort of comparable type.
+
+
+class SynchronizedConsumerBackend(ConsumerBackend[TStream, TOffset, TValue]):
+    def __init__(
+        self,
+        backend: ConsumerBackend[TStream, TOffset, TValue],
+        commit_log_consumer_backend: ConsumerBackend[
+            # NOTE: These TStream and TOffset values for the commit log
+            # consumer do not need to match, but this simplifies things a bit
+            # so that we don't have to declare two additional type variables.
+            TStream,
+            TOffset,
+            CommitData[TStream, TOffset],
+        ],
+        remote_consumer_groups: Set[ConsumerGroup],
+        get_commit_log_streams_for_topic: Callable[[Topic[TStream]], Sequence[TStream]],
```

Comment:
This makes the assumption that for a specific topic definition, we can pinpoint the exact set of topic and partitions where the offsets for that topic definition are stored. The alternative approach here would be that a topic definition just maps to another topic definition, but this would require that the commit log consumer becomes a balanced consumer (in a consumer group with only one member) for better or for worse. Not sure which.